src/comm/TcpAcceptor.cc

```diff
@@ -271,33 +272,41 @@ Comm::TcpAcceptor::acceptOne()
 
     /* Accept a new connection */
     ConnectionPointer newConnDetails = new Connection();
-    const Comm::Flag flag = oldAccept(newConnDetails);
-
-    if (flag == Comm::COMM_ERROR) {
-        // A non-recoverable error; notify the caller */
-        debugs(5, 5, "non-recoverable error:" << status() << " handler Subscription: " << theCallSub);
-        if (intendedForUserConnections())
-            logAcceptError(newConnDetails);
-        notify(flag, newConnDetails);
-        // XXX: not under async job call protections
-        mustStop("Listener socket closed");
+    try {
+        if (acceptInto(newConnDetails)) {
+            Assure(newConnDetails->isOpen());
+            CallBack(newConnDetails, [&] {
+                debugs(5, 5, "Listener: " << conn <<
+                       " accepted new connection " << newConnDetails <<
+                       " handler Subscription: " << theCallSub);
+                notify(Comm::OK, newConnDetails);
+            });
+        } else {
+            debugs(5, 5, "try later: " << conn << " handler Subscription: " << theCallSub);
+            newConnDetails->close(); // paranoid manual closure (and may already be closed)
+        }
+
+        SetSelect(conn->fd, COMM_SELECT_READ, doAccept, this, 0);
         return;
+    } catch (...) {
+        const auto debugLevel = intendedForUserConnections() ? DBG_CRITICAL : 3;
+        debugs(5, debugLevel, "ERROR: Stopped accepting connections:" <<
+               Debug::Extra << "error: " << CurrentException);
     }
 
-    if (flag == Comm::NOMESSAGE) {
-        /* register interest again */
-        debugs(5, 5, "try later: " << conn << " handler Subscription: " << theCallSub);
-    } else {
-        // TODO: When ALE, MasterXaction merge, use them or ClientConn instead.
-        CallBack(newConnDetails, [&] {
-            debugs(5, 5, "Listener: " << conn <<
-                   " accepted new connection " << newConnDetails <<
-                   " handler Subscription: " << theCallSub);
-            notify(flag, newConnDetails);
-        });
-    }
+    if (intendedForUserConnections())
+        logAcceptError(newConnDetails);
+
+    // do not expose subscribers to a "usable" descriptor of a failed connection
+    newConnDetails->close(); // may already be closed
 
-    SetSelect(conn->fd, COMM_SELECT_READ, doAccept, this, 0);
+    CallBack(newConnDetails, [&] {
+        notify(Comm::COMM_ERROR, newConnDetails);
+    });
+
+    // XXX: Not under AsyncJob call protections but, if placed there, may cause
+    // problems like making the corresponding HttpSockets entry (if any) stale.
+    mustStop("unrecoverable accept failure");
```

Comment:
I actually addressed this old XXX in this branch by switching from mustStop() to deleteThis(). deleteThis() does work as intended here, closing the listening socket and destroying the TcpAcceptor job, but I reverted those branch changes because:

1. If this TcpAcceptor object is tied to an HttpSockets FD entry, that entry will become stale. That might be OK, but it would be better to avoid creating this problem anyway (probably by carefully removing HttpSockets!).
2. A later clientConnectionsClose() call will mislead about closing (an already closed!) listening socket. It is not clear to me whether there is other code that also assumes a non-nil `HttpPortList[i]->listenConn` implies an open listening connection.
3. We should avoid adding more/new deleteThis() calls.
4. Properly fixing this old XXX (i.e. addressing the above and similar concerns) can be viewed as being outside this PR scope. It should probably be done in several PRs.

For now, Squid will continue to log this (at debug level 5) after unrecoverable accept errors:

```
AsyncJob.cc(77) mustStop: Comm::TcpAcceptor will STALL, reason: unrecoverable accept failure
```

... and the listening queue may grow (except in those SMP environments where other workers can still accept connections queued at the same port/descriptor, I guess).

Reply:
This XXX will be resolved by PR #870 when it completes.